### PR TITLE
EOS-18713: Reset password for S3 account, IAM user and Delete s3 account

### PR DESCRIPTION
**Pull request by rohitkolapkar**
# Backend

## Problem Statement
<pre>
Story Ref (if any):  
</pre>
[EOS-18713](https://jts.seagate.com/browse/EOS-18713)

## Unit testing on RPM done
<pre>
Yes
</pre>
## Problem Description
<pre>
See the ticket for details.
</pre>
## Solution
<pre>
1. Removed creating S3 user login profile before setting up the new password, CSM admin and users with mange role are 
now able to reset S3 accounts password and S3 account user is also able to reset his own password.
The same logic is also removed from IAMUser password reset.
2. Created IAMUserPatch schema,S3 account user is now able to Reset IAM user's password.
3. added _s3_session field, S3 account is now getting deleted successfully.
</pre>

## Unit Test Cases
<pre>
1. CSM admin and users with mange role should be able to reset S3 accounts password 
2. S3 account user should be able to reset his own password
3. S3 account user should be able to Reset IAM user's password
4. S3 account should get deleted.
</pre>

Signed-off-by: Rohit Kolapkar

### CODE REVIEW

**Comment by cortx-admin** [comment]
Can one of the admins verify this patch?

**Comment by rohitkolapkar** [comment]
@avrnv @pawankumarsrivastava Please review it.

File: csm/core/controllers/s3/iam_users.py [hunk 1]
```diff
@@ -118,9 +118,11 @@ async def patch(self):
         """
         Log.debug(f"Handling  IAM USER patch request."
                   f" user_id: {self.request.session.credentials.user_id}")
+        user_name = self.request.match_info["user_name"]
+        req_ob=await self.request.json()
         try:
             schema = IamUserCreateSchema()
-            patch_body = schema.load(await self.request.json(), unknown='EXCLUDE')
+            patch_body = schema.load({"user_name": user_name,"password":req_ob['password']}, unknown='EXCLUDE')
```
Comment (kupranay) [inline csm/core/controllers/s3/iam_users.py:125]: @rohitkolapkar  Please try with this code issue
schema = IamUserCreateSchema()
        Log.debug(f"Handling create IAM USER post request."
                  f" user_id: {self.request.session.credentials.user_id}")
        **try:
            **body = await self.request.json()
            request_data = schema.load(body, unknown='EXCLUDE')**
        except ValidationError as val_err:
            raise InvalidRequest(message_args=f"Invalid request body: {val_err}")**

File: csm/core/controllers/s3/accounts.py [hunk 2]
```diff
@@ -99,6 +99,7 @@ async def delete(self):
         """Calling Stats Get Method"""
         Log.debug(f"Handling s3 accounts delete request."
                   f" user_id: {self.request.session.credentials.user_id}")
+        self._s3_session = self.request.session.credentials
```
Comment (kupranay) [inline csm/core/controllers/s3/accounts.py:102]: i can see the same code is working with SMS1 release. What is the difference ?

**Review by kupranay — CHANGES_REQUESTED** [review]

**Review by avrnv — CHANGES_REQUESTED** [review]
In general, please, give the PR a proper name, remove the screenshot from the description, it's not informative.
In particular, I've got inputs on both of your fixes, need time to articulate them. Leaving this comment for the PR not to get merged until our discussion.

File: csm/core/controllers/s3/iam_users.py (hunk 1 above)
Comment (rohitkolapkar) [reply csm/core/controllers/s3/iam_users.py:125]: @kupranay It is not working, IamUserCreateSchema() requires **user_name** and **password** fields. and in self.request.json(), i am getting only **password** field. That's why I am extracting user_name from URI path.

File: csm/core/services/s3/accounts.py [hunk 3]
```diff
@@ -178,16 +178,8 @@ async def patch_account(self, account_name: str, password: str = None,
...
-            # We will try to create login profile in case it doesn't exist
-            new_profile = await client.create_account_login_profile(account_name, password)
-            if isinstance(new_profile, IamError) and \
-                    new_profile.error_code != IamErrors.EntityAlreadyExists:
-                self._handle_error(new_profile)
-
-            if isinstance(new_profile, IamError):
-                # Profile already exists, we need to set new passord
-                Log.debug(f"Update Login Profile for account {account_name}")
-                new_profile = await client.update_account_login_profile(account_name,
```
Comment (avrnv) [inline csm/core/services/s3/accounts.py:190]: Please, apply the same logic to the IAM user PATCH request, i.e. remove creating IAM user login profile before setting the password.

**Review by avrnv — CHANGES_REQUESTED** [review]

File: csm/core/controllers/s3/accounts.py (hunk 2 above)
Comment (rohitkolapkar) [reply csm/core/controllers/s3/accounts.py]: @kupranay In the latest commit, added _s3_session to its correct place as suggested by @avrnv

File: csm/core/services/s3/accounts.py (hunk 3 above)
Comment (avrnv) [reply csm/core/services/s3/accounts.py:190]: Please, do not resolve conversations yourself, it's reviewers job.

**Review by avrnv — APPROVED** [review]

**Review by pawankumarsrivastava — APPROVED** [review]

**Review by kupranay — APPROVED** [review]

**Comment by kupranay** [comment]
ok to test

**Comment by kupranay** [comment]
ok to test

**Comment by kupranay** [comment]
run mini-provisioning

**Pull request merged**